# Setting up data import

In [0]:
!pip install --upgrade kaggle

Requirement already up-to-date: kaggle in /usr/local/lib/python3.6/dist-packages (1.5.6)


In [0]:
import os
os.environ['KAGGLE_USERNAME'] = "skanderborgi" # username from the json file
os.environ['KAGGLE_KEY'] = "1c342e34d4e302ada65ba18030659145" # key from the json file
!kaggle competitions download -c santander-customer-satisfaction

test.csv.zip: Skipping, found more recently modified local copy (use --force to force download)
train.csv.zip: Skipping, found more recently modified local copy (use --force to force download)
sample_submission.csv: Skipping, found more recently modified local copy (use --force to force download)


In [0]:
cd /content/

/content


In [0]:
ls

sample_data/           test.csv      train.csv
sample_submission.csv  test.csv.zip  train.csv.zip


In [141]:
!unzip "train.csv.zip"
!unzip "test.csv.zip"

Archive:  train.csv.zip
replace train.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: n
Archive:  test.csv.zip
replace test.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: n


In [142]:
ls

sample_data/           test.csv      train.csv
sample_submission.csv  test.csv.zip  train.csv.zip


# Data exploration

In [0]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [144]:
train=pd.read_csv("train.csv",index_col=0)
train.head()

var3  var15  ...          var38  TARGET
ID               ...                       
1      2     23  ...   39205.170000       0
3      2     34  ...   49278.030000       0
4      2     23  ...   67333.770000       0
8      2     37  ...   64007.970000       0
10     2     39  ...  117310.979016       0

[5 rows x 370 columns]

In [0]:
test=pd.read_csv("test.csv",index_col=0)

In [0]:
train=train.dropna(axis=0)

In [0]:
X=train.iloc[:,:-1]
y=train['TARGET']

In [148]:
X.describe()

var3         var15  ...  saldo_medio_var44_ult3         var38
count   76020.000000  76020.000000  ...            76020.000000  7.602000e+04
mean    -1523.199277     33.212865  ...               56.614351  1.172358e+05
std     39033.462364     12.956486  ...             2852.579397  1.826646e+05
min   -999999.000000      5.000000  ...                0.000000  5.163750e+03
25%         2.000000     23.000000  ...                0.000000  6.787061e+04
50%         2.000000     28.000000  ...                0.000000  1.064092e+05
75%         2.000000     40.000000  ...                0.000000  1.187563e+05
max       238.000000    105.000000  ...           397884.300000  2.203474e+07

[8 rows x 369 columns]

we can see by looking at the mean that there is a huge difference in the values of the features, we need to scale them

## Feature selection

#### Removing features with low variance
It removes all features whose variance doesn’t meet some threshold. By default, it removes all zero-variance features.

In [149]:
from sklearn.feature_selection import VarianceThreshold
sel_variance_threshold = VarianceThreshold() 
X = sel_variance_threshold.fit_transform(X)
X.shape

(76020, 335)

#### Checking our target

In [150]:
y.describe()

count    76020.000000
mean         0.039569
std          0.194945
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          1.000000
Name: TARGET, dtype: float64

In [151]:
y.value_counts()

0    73012
1     3008
Name: TARGET, dtype: int64

=> We can see that the target variables are imbalanced : 73012 for 0 and only 3008 for 1

## How to deal with this problem?
#### We have two approaches :
We can use **SMOTE**  algorithm (Synthetic Minority Over-Sampling Technique):
It's a combination of oversampling and undersampling, but instead of replicating minority class, we construct new minority class data instance via an algorithm.



In [152]:
from collections import Counter
from sklearn.datasets import make_classification
from imblearn.over_sampling import SMOTE 

smt = SMOTE()
Xsmt, ysmt = smt.fit_sample(X, y)

/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)


In [153]:
np.bincount(ysmt) #Now they both have same value

array([73012, 73012])

## Feature scaling

In [0]:
  from sklearn import preprocessing

scaler = preprocessing.StandardScaler()
Xsmt = scaler.fit_transform(Xsmt)

In [155]:
Xsmt

array([[ 0.03247351, -1.09176273, -0.05697819, ..., -0.01504036,
        -0.01632659, -0.46646954],
       [ 0.03247351, -0.18537064, -0.05697819, ..., -0.01504036,
        -0.01632659, -0.39889018],
       [ 0.03247351, -1.09176273, -0.05697819, ..., -0.01504036,
        -0.01632659, -0.27775323],
       ...,
       [ 0.03247351,  0.12694937, -0.05284696, ..., -0.01504036,
        -0.01632659, -0.43213196],
       [ 0.03247351, -0.26776992, -0.05697819, ..., -0.01504036,
        -0.01632659,  0.05754659],
       [ 0.03247351,  0.73808348, -0.05697819, ..., -0.01504036,
        -0.01632659, -0.06760522]])